src/reducers/modal.reducer.ts

```diff
@@ -1,17 +1,53 @@
+import { ButtonProps } from '@material-ui/core';
 import { Modal } from '../actions/modal.actions';
 
+export interface ModalButton extends ButtonProps {
+  text: string,
+  value?: any,
+}
+
+export type ModalResponse = {
+  button: ModalButton,
+  index: number,
+} | null;
+
 export interface ModalState {
+  buttons: ModalButton[],
   message: string,
   open: boolean,
   title: string,
 }
 
 export const modalInitialState: ModalState = {
+  buttons: [],
   open: false,
   message: '',
   title: '',
 };
 
+export type ResolverStore = {
+  [key: string]: (response: ModalResponse) => void
+};
+
+export const modalResolverStore: ResolverStore = {};
+
+export function modalResolverHandler() {
+  return (next: (any: any) => any) => (action: any) => {
+    if (action.type === 'MODAL_ALERT') {
+      const { payload, resolve } = (action as Modal.Actions.Alert);
+      modalResolverStore[payload.title] = resolve;
+    }
+    if (action.type === 'MODAL_CLOSE') {
+      const payload = (action as Modal.Actions.Close).payload;
+      if (typeof modalResolverStore[payload.title] === 'function') {
+        modalResolverStore[payload.title](payload.response);
+        delete modalResolverStore[payload.title];
+      }
+    }
+    return next(action);
+  };
+}
+
```

Comment:
Is this a workaround to avoid saving functions in the Redux store? It looks like this is essentially doing that anyways, just in a hidden store.

What do you think about going ahead and putting the resolve functions in the main store? If we're already essentially doing the same thing I'd rather just keep it simple if we can. I know it's not recommended in Redux, but it's not strictly against the rules. Modals are such a temporary part of the UI that I don't think it would actually hurt anything.

If we do need to do it this way, I'd drop some comments in there at least.

Reply:
Putting them in the main store would break serialization. In theory. This pattern is from: https://stackoverflow.com/questions/35325195/should-i-store-function-references-in-redux-store#:~:text=No%2C%20you%20should%20not%20store,of%20hotkeys%20to%20their%20actionCreatorFuncNames.